pkg/controller/kubelet-config/kubelet_config_bootstrap.go

```diff
@@ -0,0 +1,165 @@
+package kubeletconfig
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+
+	configv1 "github.com/openshift/api/config/v1"
+	mcfgv1 "github.com/openshift/machine-config-operator/pkg/apis/machineconfiguration.openshift.io/v1"
+	ctrlcommon "github.com/openshift/machine-config-operator/pkg/controller/common"
+	"github.com/openshift/machine-config-operator/pkg/version"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	kubeletCfgMasterLabel = "pools.operator.machineconfiguration.openshift.io/master"
+	kubeletCfgWorkerLabel = "pools.operator.machineconfiguration.openshift.io/worker"
+)
+
+// RunKubeletBootstrap generates MachineConfig objects for mcpPools that would have been generated by syncKubeletConfig
+func RunKubeletBootstrap(templateDir string, kubeletConfigs []*mcfgv1.KubeletConfig, controllerConfig *mcfgv1.ControllerConfig, features *configv1.FeatureGate, mcpPools []*mcfgv1.MachineConfigPool) ([]*mcfgv1.MachineConfig, error) {
+	generatedManagedKeys := make(map[string]string)
+	generatedMCs := make(map[string]*mcfgv1.MachineConfig)
+	// Validate the KubeletConfig CR if exists
+	for _, kubeletConfig := range kubeletConfigs {
+		if kubeletConfig == nil {
+			continue
+		}
+		if err := validateUserKubeletConfig(kubeletConfig); err != nil {
+			return nil, newForgetError(err)
+		}
+	}
+
+	userDefinedSystemReserved := make(map[string]string)
+	var res []*mcfgv1.MachineConfig
+	for _, kubeletConfig := range kubeletConfigs {
+		if kubeletConfig == nil || kubeletConfig.Spec == (mcfgv1.KubeletConfigSpec{}) {
+			continue
+		}
+		for _, pool := range mcpPools {
```

Comment:
I was just testing this with the work I've been doing in #2687 and realised that the label matching part of a KubeletConfig is not handled during the bootstrap, leading to day 1 vs day 2 discrepancies.

I added the following to my branch locally and now my testing passes, I think this is correct as I've copied the logic from the controller
```suggestion
		selector, err := metav1.LabelSelectorAsSelector(kubeletConfig.Spec.MachineConfigPoolSelector)
		if err != nil {
			return nil, fmt.Errorf("invalid label selector: %v", err)
		}

		for _, pool := range mcpPools {
			// If a pool with a nil or empty selector creeps in, it should match nothing, not everything.
			if selector.Empty() || !selector.Matches(labels.Set(pool.Labels)) {
				continue
			}
```
https://github.com/openshift/machine-config-operator/blob/657590c4b7b84902b100761b2385d93e5d04522c/pkg/controller/kubelet-config/kubelet_config_controller.go#L791-L801

As this is today, it will always generate the kubelet config for both master and worker pools, independent of the selector applied to the kubelet config.

Reply:
@JoelSpeed Thanks. I added your changes. Do we still need the following lines? https://github.com/openshift/machine-config-operator/blob/6c88f51174e409b17f340505bcafd8a19bc345be/pkg/controller/kubelet-config/kubelet_config_bootstrap.go#L52-L57

Reply:
I don't know where those particular lines came from, but as far as I can tell no. This doesn't appear to be logic that matches the day 2, so could cause issues. That said, we don't have any non master/worker MCPs in an install by default, so it may not be a problem for now?

I'd defer to someone more knowledgable in this area, eg @yuqi-zhang for that call

Reply:
Sorry didn't see this earlier. I think those lines should not be in, since the user can conceivably modify their pool labelling during install time. We are probably fine with them using a non-default label during install since we are just doing general matching anyways.